feat(replays): Add buffered consumer implementation

cmanallen opened the pull request:
Edit: Out of time for now.  Converting to a draft.  Will pick this up at a later date.

---

TODO description...

Partially addresses the DACI: https://www.notion.so/sentry/DACI-Session-Replay-Recording-Consumer-Stability-and-Performance-Improvements-19e8b10e4b5d80a192a1ecd46f13eebb

How it works:
- As messages come in they are processed and their processed results are stored on a queue.
- When the queue fills up the processed messages are flushed.
    - Flushing involves committing data to GCS, ClickHouse, BigQuery, DataDog.
    - Anything I/O related.
- Flushing happens in a thread-pool.

This closely mirrors current production behavior except processing is no longer done in the thread-pool.

This PR also introduces a new RunTime abstraction for managing state changes in the consumer.  Which I will document in a Notion doc.

cmanallen committed:
Add buffered consumer runtime implementation

cmanallen committed:
Begin adding process logic

cmanallen committed:
Add tracing to each component of recording processing

cmanallen committed:
Delete unused function

cmanallen committed:
Report size metrics

cmanallen committed:
Merge branch 'cmanallen/replays-improve-tracing' into cmanallen/replays-consumer-separate-processing-from-io

cmanallen committed:
Separate IO from processing

cmanallen committed:
Add explicit return

cmanallen committed:
Merge branch 'cmanallen/replays-consumer-separate-processing-from-io' into cmanallen/replays-optimize-consumer

cmanallen committed:
Add buffer managers

cmanallen committed:
Write FilePart rows and adopt a new subscription model

cmanallen committed:
Add unit tests

cmanallen committed:
Add contextual errors

cmanallen committed:
Misc test updates

cmanallen committed:
Fully separate compute and io within the recording consumer

cmanallen committed:
Configure max workers

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-separated-compute-and-io

cmanallen committed:
Remove conditional branch as its moved further up the hierarchy

cmanallen committed:
Merge branch 'cmanallen/replays-add-separated-compute-and-io' into cmanallen/replays-optimize-consumer

cmanallen committed:
Use context manager

cmanallen committed:
Simplify buffer flushing (for now)

cmanallen committed:
Merge branch 'master' into cmanallen/replays-optimize-consumer

cmanallen committed:
Update tracing logic

cmanallen committed:
Soften flag requirements and minor fixes

cmanallen committed:
Remove buffer managers module

cmanallen committed:
Test clean up

cmanallen committed:
Fix unit test

cmanallen committed:
Add explicit return

cmanallen committed:
Fix typing

cmanallen committed:
Remove unused option

cmanallen committed:
Reset dom_index module to align with master

cmanallen committed:
Update buffering run-time coverage

cmanallen committed:
Update test ordering

cmanallen committed:
Add offset committing test

cmanallen committed:
Add docs

cmanallen committed:
More docstring fixes

cmanallen committed:
Add typing to flags and factory module

cmanallen committed:
Adopt buffered strategy in callsite

cmanallen committed:
Add script for mocking recordings

cmanallen committed:
Add handling for appending offsets when the message is not buffered

cmanallen committed:
Add commit coverage

cmanallen committed:
Assert messages are committed regardless of if they're appended to the buffer

cmanallen committed:
Fix typing

cmanallen committed:
Docstrings

cmanallen committed:
More docstrings

cmanallen committed:
Yet more docstrings

fpacifici commented on src/sentry/replays/consumers/buffered/factory.py:
```diff
@@ -0,0 +1,32 @@
+            "max_workers": max_workers,
+        }
+
+    def create_with_partitions(
+        self,
+        commit: ArroyoCommit,
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        return PlatformStrategy(commit=commit, flags=self.flags, runtime=recording_runtime)
```
Wouldn't it be considerably simpler to model this consumer as a sequence of these Arroyo operators:
- Batch the messages with the Reduce primitive (https://getsentry.github.io/arroyo/strategies/reduce.html#reduce-fold)
- Run Task in either multi threading or multi processing (https://getsentry.github.io/arroyo/strategies/reduce.html#reduce-fold) to perform the parsing and the `commit_recording_message` ?

Modeling the system this way would:
- allow the parallelism either via processes and threads without application logic changes
- guarantee a pipeline approach that allows the batching step to keep batching new messages while the worker thread performs its work.
- hide offset management entirely from the application code.

fpacifici commented on src/sentry/replays/consumers/buffered/lib.py:
```diff
@@ -0,0 +1,143 @@
+    buffer: list[Item]
+    can_flush: Callable[["Model[Item]"], bool]
+    do_flush: Callable[["Model[Item]"], None]
+    offsets: MutableMapping[Partition, int]
```
I think this is problematic, it complicates the architecture and makes mistakes easier.
Arroyo provides offset management out of the box to hide it almost entirely from the application. "almost" means that the user needs to pick how often to commit and to add a Commit strategy at the end of the consumer.

Other than that messages carry offsets, transformed messages carry the same offsets and batches carry watermarks. These are just managed by the apis. If the arroyo primitives are used, the developer never deals with offsets. 

Here we are resurfacing them and we manage offsets and watermarks explicitly. Which means we would be doing the same thing twice.

fpacifici commented on src/sentry/replays/consumers/buffered/consumer.py:
```diff
@@ -0,0 +1,128 @@
+
+        self.__last_flushed_at = time.time()
+
+    def can_flush(self, model: Model[ProcessedRecordingMessage]) -> bool:
+        # TODO: time.time is stateful and hard to test. We should enable the RunTime to perform
+        #       managed effects so we can properly test this behavior.
+        return (
+            len(model.buffer) >= self.__max_buffer_length
+            or (time.time() - self.__max_buffer_wait) >= self.__last_flushed_at
+        )
```
Arroyo primitives manage these kind of concerns for you (when to flush a batch for example). Are you sure about the idea of pushing them into the product code instead ?

fpacifici reviewed (commented):
I would advise against building a consumer with this abstraction and make it work on top of Arroyo.
- Arroyo provides an abstraction level that is similar to the one provided by your PlatformStrategy. The two, though, go in fundamentally different directions. Building one on top of the other seems to create a quite complex architecture. It is quite hard to intuitively tell whether the system guarantees at least once delivery.
- We will not be able to move parts of the processing on a multi process pool in case it is needed for scale as that qould require using the RunTask abstraction.
- As we discussed in the past. I believe the dataflow model (even the small subset provided by arroyo) makes this kind of applications simpler to understand due to the sequential nature of the pipeline.

cmanallen committed:
Implement declarative effect management

cmanallen committed:
Merge branch 'master' into cmanallen/replays-optimize-consumer

cmanallen committed:
Move offset management into the runtime

cmanallen committed:
Fix typing

cmanallen committed:
Remove comments on offsets

cmanallen committed:
Remove none-type messages

cmanallen commented:
@fpacifici Thanks for the feedback! Let me preface by saying when I opened this PR I did so for three reasons:

1. To demonstrate the behaviors I outlined in my streaming platform proposal so you would have something tangible to reference.
2. To explore new ideas as part of a research project. The goal is exploration so using Arroyo primitives would have gone against my goals.
3. I'm trying to solve a real problem from a pre-determined starting point with constraints on what can change and how fast.

So there are practical concerns intermixed with educational and exploratory concerns.  I've since dropped the educational component and simplified what I'm doing in the PR with this morning's commits (it mostly doesn't impact your review -- but it did simplify the consumer implementation so it might be worth a second look).  Hopefully that explains why I made some of the choices I did.

### Concerns Around Committing

I agree completely.  They were a pain to manage and it was never necessary to manage them in application code (for my use case at least).  I moved the offset handling into the RunTime where they're now managed at the platform level.

### Why Not Use Arroyo

Its a research project so we're trying something new.  But, critically, this is Arroyo.  Its not a full implementation but it is an Arroyo strategy.  That means I can prefix the step with any number of streaming primitives and I _could_ suffix the step with any number of streaming primitives (could being the operative word because I'm currently hard coding the commit step as the next step -- that's an easy fix so I'm ignoring this oversight).

This RunTime strategy is a generalization of all Arroyo strategies so its not surprising that specialized strategies exist that can solve components of this pipeline.  Research aside I do want this to go to production and I'll mention why at the bottom.

---

So the three concerns I'm keying in on are:

1. allow the parallelism either via processes and threads without application logic changes
2. guarantee a pipeline approach that allows the batching step to keep batching new messages while the worker thread performs its work.
3. hide offset management entirely from the application code.

Three is gone; I've removed it.  One is partially solved I think.  We can prefix the RunTime step with a multi-threading/processing step.  I'm not sure if the RunTime can be embedded into those steps so that may be a shortcoming.  That's an area I could look into if this was ever an important component of Arroyo.

Two is not solved as far as I'm aware.  I wrote the `Buffer` strategy in Arroyo and the `Reduce` strategy implements the `Buffer` strategy.  I'm not aware how those strategies flush their buffers in a worker thread.  As far as I know they block the main thread.  But if that's not the case and, there is some platform magic happening, then I don't see why the RunTime strategy couldn't also have the benefit of flushing off the main thread.

---

I want this to go to production; why?  tl;dr I can now unit-test my consumer end-to-end.

Testing is a huge concern for me.  I've refactored this consumer in the past and its led to production outages (using [Arroyo streaming primitives](https://github.com/getsentry/sentry/blob/master/src/sentry/replays/consumers/recording_buffered.py#L137-L147) as it happens -- which I don't blame).

There are minor things that Arroyo does that can make testing more difficult.  For example [in the Reduce strategy](https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/reduce.py#L48) we call `time.time()`.  That can make unit-testing difficult.  You have to mock or just not test certain behaviors which is not ideal.  However, there's a larger problem that I'm trying to address and that's the difficulty in testing how state is threaded through a consumer.  Arroyo does not provide any facilities for this and my PR did not have any until this morning.

One of the benefits of managing the state machine in the way I have is that I can intercept the commands being issued by my application and rewrite them.  You can see [in the MockRunTime class](https://github.com/getsentry/sentry/pull/85356/files#diff-0bed0de9414268ddd5688500265828a379f1bf41d5c84a706a752e79831dc15aR18-R26) I'm using coroutines to [rewrite commands in my test suite](https://github.com/getsentry/sentry/pull/85356/files#diff-39b5ba3bc6f85c1e77efedf3e9172c8fc354617b3374291a80fff71cdc0f3292R78-R88).  This gives me a lot of insight into what the application is doing and the ability to redirect behavior in a way that does not require monkey patching.  I can deterministically simulate all possible states and assert the outcome very cheaply.

My implementation isn't perfect (I haven't abstracted all state yet) but there's already been a significant uplift in what I'm capable of asserting about my software.

There are other reasons but this has already gotten too long so I'll leave it there.  Let me know if I addressed your concerns well enough and thanks again for taking the time to review this!

cmanallen committed:
Move offsets out of runtime and into platform strategy. Add support for arbitrary next step

cmanallen committed:
Fix typing

cmanallen committed:
Docstrings and renames

fpacifici commented:
> I want this to go to production; why? tl;dr I can now unit-test my consumer end-to-end.

This is the main reason that prompted me to express my concerns on this PR. 
Replays is a high tier production service. Somebody will have to be on-call for this and guarantee its stability and scale to meet the SLOs.

Arroyo primitives have been designed over time to ensuring certain failover guarantees, provide the observability needed to troubleshoot production issues and prevent certain patterns. 
In practice SRE (soon platform) can be on call for consumers they did not write because it is reasonably possible to troubleshooting their issues from what arroyo provides. This works as long as consumers follow the patterns provided by Arroyo. 
A considerably different architecture will break those assumptions unless the failure modes and observability needs are experienced before in other environments. 

Consider one of the following:
- take the pager for it. This can be arranged with SRE. 
- test it in the sandbox environment under synthetic generated load 
- deploy this in S4S only for enough time to understand how to manage this consumer.

cmanallen commented:
@fpacifici 

> take the pager for it

😧 I'll implement this strictly in Arroyo primitives.  Thank you for your time.

Jokes aside this is a great point "Replays is a high tier production service. Somebody will have to be on-call for this and guarantee its stability and scale to meet the SLOs.".

The consumer as it exists now could absolutely be implemented in Arroyo primitives.  I'm implementing it differently because in the future I want to have state which persists between flushes and I wanted really strong unit testing.  However, I can accomplish these goals through other means (using a global, using slower integration tests, using monkey-patching, etc.).

If you tell me that a pager is required for this to live in prod then I can abandon this RunTime implementation and use the Reduce step.  I don't want to be a burden on SRE and I can't take on the responsibility of a pager.

> test it in the sandbox environment under synthetic generated load
> deploy this in S4S only for enough time to understand how to manage this consumer.

Notion doc for the sandbox? And testing in S4S first is my intention.

cmanallen committed:
Update error handling and documentation

cmanallen committed:
Rename to sandbox

cmanallen committed:
Update test coverage

cmanallen committed:
Remove unnecessary lambda

cmanallen committed:
Improve coveragE

cmanallen committed:
Sketch selective retry

fpacifici commented:
> I don't want to be a burden on SRE and I can't take on the responsibility of a pager.

Thank, that is appreciated. Let me expand a bit on this problem to ensure we are on the same page.
You are right. The api allows you to build this. I believe it should be a bit more opinionated in what the application can do in order to be able to operate the system.
The issue is not that this approach cannot be made reliable or scalable. I am sure it can. Though that would mean for the platform team (which will take over consumers from SREs) maintaining two different architectures with different failure modes. 
I believe that would be a good investment if there was a strong product or technical gain in moving in that direction and if we had a plan to converge to one architecture. That's what we are trying to do with the streaming platform: the technical reason is to raise the abstraction level for product teams writing applications to reduce the infrastructure concerns they have to deal with. The plan is also to fully move all consumers there. 

You raised some legitimate concerns that we could improve on, specifically unit tests. 
I would advise we try to fix that and ensure the concern is taken into account in the platform runtime. So I would encourage you to list the other limitations you ran into when trying to build your consumer using the Arroyo abstractions so we can triage and prioritize them.
